### PR TITLE
Fixed a thread safety issue that could cause application to crash in …

### PR DESCRIPTION
**Pull request by zoewangg**
…the edge case in AWS CRT HTTP client

## Motivation and Context
https://github.com/awslabs/aws-crt-java/issues/748#issuecomment-1899288531
In #4833, we used AtomicCounter to make sure we don't invoke stream.incrementWindow *after* the stream is closed. However, this is not sufficient since the stream could get closed right after the check.

## Modifications
Used a lock to guard against operations using stream, connection

## Testing
It's an edge case and not reproducible locally, so no tests are added as part of this PR.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Review by TingDaoK — APPROVED** [review]

**Comment by joviegas** [comment]
> ## [Quality Gate Failed](https://sonarcloud.io/dashboard?id=aws_aws-sdk-java-v2&pullRequest=4839) **Quality Gate failed**
> 
> Failed conditions
> 
> [59.1% Coverage on New Code](https://sonarcloud.io/component_measures?id=aws_aws-sdk-java-v2&pullRequest=4839&metric=new_coverage&view=list) (required ≥ 80%)
> 
> [See analysis details on SonarCloud](https://sonarcloud.io/dashboard?id=aws_aws-sdk-java-v2&pullRequest=4839)

Is this false positive ?

File: http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/response/ResponseHandlerHelper.java [hunk 1]
```diff
@@ -57,20 +59,34 @@ public void onResponseHeaders(HttpStream stream, int responseStatusCode, int hea
      * Release the connection back to the pool so that it can be reused.
      */
     public void releaseConnection(HttpStream stream) {
-        if (connectionClosed.compareAndSet(false, true)) {
-            connection.close();
-            stream.close();
+        synchronized (lock) {
```
Comment (joviegas) [inline http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/response/ResponseHandlerHelper.java:62]: Is it possible to add Junit test cases for this ?

Comment (zoewangg) [reply http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/response/ResponseHandlerHelper.java:61]: Yeah, added tests

File: http-clients/aws-crt-client/pom.xml [hunk 2]
```diff
@@ -105,7 +105,7 @@
         </dependency>
         <dependency>
             <groupId>org.mockito</groupId>
-            <artifactId>mockito-core</artifactId>
+            <artifactId>mockito-inline</artifactId>
```
Comment (zoewangg) [inline http-clients/aws-crt-client/pom.xml:108]: This allows us to mock final classes/methods

File: http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/response/ResponseHandlerHelper.java [hunk 3]
```diff
@@ -57,20 +58,34 @@ public void onResponseHeaders(HttpStream stream, int responseStatusCode, int hea
...
+            if (!connectionClosed) {
+                connectionClosed = true;
+                connection.close();
+                stream.close();
+            }
+        }
+    }
+
+    public void incrementWindow(HttpStream stream, int windowSize) {
```
Comment (joviegas) [inline http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/response/ResponseHandlerHelper.java:70]: Could you please confirm if we have testes and confirmed that  there is not latency because of this synchronized check in multi threaded environment  with multiple requests going on at the same time?

**Review by joviegas — APPROVED** [review]

File: http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/response/ResponseHandlerHelper.java (hunk 3 above)
Comment (zoewangg) [reply http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/response/ResponseHandlerHelper.java:70]: I don't anticipate any latency change as part of this change since most of times, there should only be one thread accessing those methods. The only exception is that when the request fails, it may be possible one thread is trying to invoke incrementWindow and the other is closing it.

I'll double check our benchmarks after today's release.
